StyleCop.Analyzers/StyleCop.Analyzers/GeneratedCodeAnalysisExtensions.cs

```diff
@@ -0,0 +1,108 @@
+    using System.Runtime.CompilerServices;
+    using System.Text.RegularExpressions;
+    using System.Threading;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    internal static class GeneratedCodeAnalysisExtensions
+    {
+        /// <summary>
+        /// The value for a SyntaxTree in the <see cref="GeneratedHeaderPresentCheck"/> table
+        /// when the document has an auto-generated header.
+        /// </summary>
+        private static readonly object GeneratedHeaderPresent = new object();
+
+        /// <summary>
+        /// The value for a SyntaxTree in the <see cref="GeneratedHeaderPresentCheck"/> table
+        /// when the document does NOT have an auto-generated header.
+        /// </summary>
+        private static readonly object GeneratedHeaderAbsent = new object();
+
+        /// <summary>
+        /// A cache of the result of computing whether a document has an auto-generated header.
+        /// </summary>
+        /// <remarks>
+        /// This allows many analyzers that run on every token in the file to avoid checking
+        /// the same state in the document repeatedly.
+        /// </remarks>
+        private static readonly ConditionalWeakTable<SyntaxTree, object> GeneratedHeaderPresentCheck = new ConditionalWeakTable<SyntaxTree, object>();
+
+        internal static bool IsGenerated(this SyntaxNodeAnalysisContext context, CancellationToken cancellationToken) => (context.Node.SyntaxTree?.IsGenerated(cancellationToken) ?? false) || (context.Node?.IsGenerated() ?? false);
+
+        internal static bool IsGenerated(this SyntaxTreeAnalysisContext context, CancellationToken cancellationToken) => context.Tree?.IsGenerated(cancellationToken) ?? false;
+
+        internal static bool IsGenerated(this SymbolAnalysisContext context, CancellationToken cancellationToken)
+        {
+            if (context.Symbol == null) return false;
+            foreach (var syntaxReference in context.Symbol.DeclaringSyntaxReferences)
+            {
+                if (syntaxReference.SyntaxTree.IsGenerated(cancellationToken)) return true;
+                var root = syntaxReference.SyntaxTree.GetRoot(cancellationToken);
+                var node = root?.FindNode(syntaxReference.Span);
+                if (node.IsGenerated()) return true;
+            }
+            return false;
+        }
+
+        internal static bool IsGenerated(this SyntaxNode node) => node.HasAttributeOnAncestorOrSelf("DebuggerNonUserCode", "GeneratedCode");
+
+        internal static bool IsGenerated(this SyntaxTree tree, CancellationToken cancellationToken) => (tree.FilePath?.IsOnGeneratedFile() ?? false) || tree.HasAutoGeneratedComment(cancellationToken);
+
+        internal static bool HasAutoGeneratedComment(this SyntaxTree tree, CancellationToken cancellationToken)
+        {
+            lock (GeneratedHeaderPresentCheck)
```

Comment:
This entire block can be shortened to the following (no `lock` required, and note that `box` is used again in a comment below).

```csharp
StrongBox<object> box = GeneratedHeaderPresentCheck.GetOrCreateValue(tree);
if (box.Value != null)
{
    return box.Value == GeneratedHeaderPresent;
}
```